prefetch.bs

```diff
@@ -252,233 +325,283 @@ A <dfn export>cross-origin prefetch IP anonymization policy</dfn> has an <dfn ex
 
-            <div class="note">This explicitly acknowledges that implementations might have additional restrictions. For instance, anonymized traffic might not be possible to some hosts, such as those that are not publicly routable and those that have <a href="https://buettner.github.io/private-prefetch-proxy/traffic-advice.html">traffic advice</a> declining private prefetch traffic.
-        1. [=redirect chain/Append=] |request| to |prefetchRecord|'s [=prefetch record/redirect chain=].
-        1. Return "`Allowed`".
+</div>
 
-    1. Let |shouldBlockNavigationResponse| be the following steps, given [=request=] |request| and [=response=] |response|:
-        1. Let |responseCOOP| be the result of [=obtaining a cross-origin opener policy=] given |response| and |request|'s [=request/reserved client=].
-        1. If |prefetchRecord|'s [=prefetch record/sandboxing flag set=] is not empty and |responseCOOP|'s [=cross-origin opener policy/value=] is not "`unsafe-none`", then return "`Blocked`".
-        1. [=redirect chain/Update the response=] for |prefetchRecord|'s [=prefetch record/redirect chain=] given |request| and |response|.
+<hr>
 
-            <div class="note">This allows [=enforcing a response's cross-origin opener policy=] to be deferred, since this has visible side effects such as queuing violation reports.</div>
-        1. Return "`Allowed`".
+<div algorithm>
+    <div class="note">This is an update of the existing <a spec=HTML>create navigation params by fetching</a> algorithm.</div>
+
+    To <dfn>create navigation params by fetching</dfn> given a [=request=] |request|, a [=session history entry=] |entry|, a [=cross-origin opener policy enforcement result=] |coopEnforcementResult|, a navigable |navigable|, a [=source snapshot params=] |sourceSnapshotParams|, a [=target snapshot params=] |targetSnapshotParams|, a string |cspNavigationType|, a [=navigation ID=] or null |navigationId|, a {{NavigationTimingType}} |navTimingType|, an optional [=opaque origin=] or null (default null) |isolationOrigin|, an optional boolean |processEarlyHints| (default true), an optional boolean |reportCOOPViolations| (default true), an optional algorithm |shouldBlockNavigationRequest|, and an optional algorithm |processNavigationResponse|, perform the following steps. If given, |shouldBlockNavigationRequest| must be an algorithm accepting a [=request=] and returning a "`Blocked`" or "`Allowed`". They return a [=navigation params=], a [=non-fetch scheme navigation params=], or null. If given, |processNavigationResponse| must be an algorithm accepting a [=request=] and [=response=].
```

Comment:
Oof, this got complex.

Some parameter declaration tweaks that might help readability a bit:

- Non-optional parameters in the order _entry_, _navigable_, _request_, _coopEnforcementResult_, _sourceSnapshotParams_, _targetSnapshotParams_, _cspNavigationType_, _navigationId_, _navTimingType_ (similar to today's HTML spec)

- Optional params can be named, i.e. declared here with something like `<dfn for="create navigation params by fetching">|reportCOOPViolations|</dfn>` and then referenced with `<i>[=create navigation params by fetching/reportCOOPViolations=]</i> set to true`.

Another option that might be worth exploring is if we could make the algorithm clearly have a "prefetch mode" and a "normal mode". E.g. instead of having fully-customizable _processNavigationResponse_, pass in a prefetch record or null, and if it's non-null, call "Update the response" directly. If this were able to collapse the 5 new optional params into 1 or 2 new ones, I think that would be a big win. Otherwise, it kind of seems like all 5 can vary independently, which is a lot of complexity for the reader to keep track of.

Reply:
Tried this. It adds a bit more bulk to "create navigation params by fetching" than I'd like, though maybe some bits can be factored back out again. (I kinda leaned toward keeping them inline at least for now, in part because some of the handwaves like "use a connection which is IP-private!" probably ought not to be hidden in some "can this request continue prefetch" predicate.)